js/perf/viewer/src/main.js

```diff
@@ -0,0 +1,192 @@
+  const perfData = [], gitRevs = [];
+
+  for (let head = await ds.head(), i = 0; head && i < MAX_PERF_HISTORY; i++) {
+    const val = head.value;
+    invariant(val instanceof Struct);
+    perfData.push(val);
+    gitRevs.push(val.nomsRevision);
+
+    const parentRef = await head.parents.first(); // TODO: how to deal with multiple parents?
+    head = parentRef ? await parentRef.targetValue(ds.database) : null;
+  }
+
+  return [perfData, gitRevs];
+}
+
+// Returns a map of URL param key to value.
+function getParams(): {[key: string]: string} {
+  // Note: this way anything after the # will end up in `params`, which is what we want.
+  const params = {};
+  const paramsIdx = location.href.indexOf('?');
+  if (paramsIdx > -1) {
+    decodeURIComponent(location.href.slice(paramsIdx + 1)).split('&').forEach(pair => {
+      const [k, v] = pair.split('=');
+      params[k] = v;
+    });
+  }
+  return params;
+}
+
+async function render() {
+  if (!chartDatasets) {
+    return;
+  }
+
+  const datasets = [];
+  for (const [testName, elapsed] of chartDatasets) {
+    const [backgroundColor, borderColor] = await genLightAndDarkColors(testName);
+    datasets.push({
+      backgroundColor,
+      borderColor,
+      borderWidth: 1,
+      data: elapsed,
+      label: testName,
+    });
+  }
+
+  // $FlowIssue: Chart is in modules, not imported, so Flow doesn't know about it.
+  new Chart(document.getElementById('chart'), {
+    type: 'line',
+    data: {
+      labels: chartLabels,
+      datasets,
+    },
+    options: {
+      scales: {
+        yAxes: [{
+          scaleLabel: {
+            display: true,
+            labelString: 'elapsed (seconds)',
+          },
+          ticks: {
+            beginAtZero: true,
+          },
+        }],
+        xAxes: [{
+          scaleLabel: {
+            display: true,
+            labelString: 'github.com/attic-labs/noms git revision',
+          },
+        }],
+      },
+    },
+  });
+}
+
+// Returns the median of numbers in `nums`.
+function median(nums: number[]): number {
+  const sorted = nums.slice();
+  sorted.sort();
+  const lenDiv2 = Math.floor(nums.length / 2);
+  let res = nums[lenDiv2];
+  if (nums.length % 2 === 0) {
+    res += nums[lenDiv2 - 1];
+    res /= 2;
+  }
+  return res;
+}
+
+// Generates a light and dark version of some color randomly (but stable) derived from `str`.
+async function genLightAndDarkColors(str: string): Promise<[string, string]> {
+  const strBuf = new window.TextEncoder().encode(str);
+  const hash = await window.crypto.subtle.digest('sha-256', strBuf);
+  invariant(hash instanceof ArrayBuffer);
+  const [r, g, b] = new Uint8Array(hash);
+  return [`rgba(${r}, ${g}, ${b}, 0.3)`, `rgb(${r}, ${g}, ${b})`];
+}
+
+// Returns the keys of `map`.
+function keys<K: Value, V: Value>(map: NomsMap<K, V>): Promise<K[]> {
+  return new Promise(res => {
+    const keys = [];
+    map.forEach((_, key) => {
+      keys.push(key);
+      return;
```

Comment:
I initially did it like this to make both eslint and flow happy:
* If I don't have the empty return, then eslint will insist that I write this like `map.forEach((_, key) => keys.push(key))`
* but then flow complains about that, because `keys.push` returns an int, but `map.forEach` only takes functions that return a `?Promise<void>`, of which this isn't.
* _but_ now eslint seems to not care, so that's weird, but I'll take it.